### ISSUE TITLE
apt-get install -y keyboard-configuration hangs on prompt

### ISSUE THREAD

**Issue by davetapley** [docker/build-push-action#504]
#### Steps to reproduce this issue

See https://github.com/JEFuller/keyboard-configuration-repro, and [cancelled build](https://github.com/JEFuller/keyboard-configuration-repro/runs/4257500841).

Clone ⬆️ and observe it ⬇️ builds just fine.
```
docker buildx build .
```

#### Expected behaviour

Can build that `Dockerfile` using `docker/build-push-action@v2`.

#### Actual behaviour

Github Action will hang indefinitely on ⬇️ until manually cancelled.
```
#6 2.766 Country of origin for the keyboard: 
```

### Configuration

* Repository URL: https://github.com/JEFuller/keyboard-configuration-repro
* Build URL: https://github.com/JEFuller/keyboard-configuration-repro/runs/4257500841

### Logs

[logs_2.zip](https://github.com/docker/build-push-action/files/7566389/logs_2.zip)

---

## Workaround

Per [here](https://askubuntu.com/a/1006187/82469) can set `ENV DEBIAN_FRONTEND=noninteractive` and it works (see [this build](https://github.com/JEFuller/keyboard-configuration-repro/runs/4257617421?check_suite_focus=true)).

That's fine, but the bug is that the workaround is **not** required when building locally.

**Comment by crazy-max** [comment]
> Per [here](https://askubuntu.com/a/1006187/82469) can set `ENV DEBIAN_FRONTEND=noninteractive` and it works (see [this build](https://github.com/JEFuller/keyboard-configuration-repro/runs/4257617421?check_suite_focus=true)).

Setting `ENV DEBIAN_FRONTEND=noninteractive` is highly discouraged otherwise interactive usage will be broken inside the container with `docker run -it ... sh`. You should set that inline in your Dockerfile like `DEBIAN_FRONTEND=noninteractive apt-get install ...`. See https://docs.docker.com/engine/faq/#why-is-debian_frontendnoninteractive-discouraged-in-dockerfiles.

> the workaround is not required when building locally.

That's because `TERM` is not set on GitHub runners, that's why the `keyboard-configuration` pkg fallsback to a teletype console as shown in your logs:

```
#6 2.402 debconf: unable to initialize frontend: Dialog
#6 2.402 debconf: (TERM is not set, so the dialog frontend is not usable.)
#6 2.402 debconf: falling back to frontend: Readline
#6 2.409 debconf: unable to initialize frontend: Readline
#6 2.409 debconf: (This frontend requires a controlling tty.)
#6 2.409 debconf: falling back to frontend: Teletype
```

So long story short you have to use `DEBIAN_FRONTEND=noninteractive apt-get install ...`. Maybe @thaJeztah are more background on that kind of issue.

**Comment by thaJeztah** [comment]
The `TERM is not set` messages are not a real error, and safe to ignore., but yes, these can be tricky; I recall I ran into issues with this when Ubuntu Focal was released, which added `tzdata` as a dependency for some packages, but it defaulted to "interactive", causing builds to fail (waiting for user-input to select a time zone);


```
Setting up tzdata (2019c-3ubuntu1) ...
debconf: unable to initialize frontend: Dialog
debconf: (TERM is not set, so the dialog frontend is not usable.)
debconf: falling back to frontend: Readline
Configuring tzdata
------------------

Please select the geographic area in which you live. Subsequent configuration
questions will narrow this down by presenting a list of cities, representing
the time zones in which they are located.

  1. Africa      4. Australia  7. Atlantic  10. Pacific  13. Etc
  2. America     5. Arctic     8. Europe    11. SystemV
  3. Antarctica  6. Asia       9. Indian    12. US
Geographic area:
```

In general, for Dockerfiles based on Debian/Ubuntu (deb-based images), I'd recommend;

- to use use `DEBIAN_FRONTEND=noninteractive apt-get install ...` inline
- OR, to use `ARG DEBIAN_FRONTEND=noninteractive` (same effect, but also works if, e.g., you're installing packages in a script, etc). Do _NOT_ use `ENV DEBIAN_FRONTEND=..` as that makes the setting "permanent" for the image (which may not be what you want if you also want to use the image interactively).
- Slightly related; use `--no-install-recommends` when installing packages to keep the number of installed packages low (i.e.; don't install what you don't need). In _some_ cases this could break your Dockerfile if you depend on one of those "recommended" packages, but it's better to be explicit, and list those packages to be installed, than to depend of them being installed as "side effect" of some other package

**Comment by davetapley** [comment]
Thanks for great replies @crazy-max, @thaJeztah 🙏🏻 

In particular @thaJeztah I had no idea that `ARG` could be used in that way, very useful.

---

I was trying to think if there is a better way to handle this, but I couldn't come up with anything.
It seems to be a well worn path, here's some related issues for people in the future:

- https://github.com/moby/moby/issues/4032
- https://github.com/microsoft/vscode-dev-containers/pull/102

**Comment by thaJeztah** [comment]
Glad to hear that helped!

> I had no idea that ARG could be used in that way, very useful.

Yes, `ARG`s are also set as environment variables during the build, which makes them a nice way to set some env vars that are only used during build, but not needed afterwards (so don't need to be in the image that's produced)
